Sources/Testing/Running/Configuration.TestFilter.swift

```diff
@@ -0,0 +1,295 @@
+
+extension Configuration.TestFilter.Kind {
+  /// Apply this test filter to a test graph and remove tests that should not be
+  /// included.
+  ///
+  /// - Parameters:
+  ///   - testGraph: The test graph to filter.
+  ///
+  /// - Returns: A copy of `testGraph` with filtered tests replaced with `nil`.
+  ///
+  /// This function provides the bulk of the implementation of
+  /// ``Configuration/TestFilter/apply(to:)``.
+  fileprivate func apply(to testGraph: Graph<String, Test?>) -> Graph<String, Test?> {
+    switch self {
+    case .unfiltered:
+      return testGraph
+    case let .precomputed(selection, membership):
+      return testGraph.mapValues { test in
+        guard let test else {
+          return nil
+        }
+        return switch membership {
+        case .including:
+          selection.contains(test) ? test : nil
+        case .excluding:
+          !selection.contains(test, inferAncestors: false) ? test : nil
+        }
+      }
+    case let .function(function, membership):
+      // When filtering by predicate function, it is necessary to determine
+      // membership AFTER resolving all tests, since we do not know what the
+      // function is going to do with the test and it needs the test instance in
+      // order to do anything useful, whereas test IDs can be constructed
+      // independently of the tests they identify.
+      //
+      // The most expedient path forward is to construct a test ID selection
+      // containing matching tests, then translate it into a new instance of
+      // TestFilter, then finally run that test filter to modify the graph.
+      let testIDs = testGraph
+        .compactMap(\.value).lazy
+        .filter(function)
+        .map(\.id)
+      let selection = Test.ID.Selection(testIDs: testIDs)
+      return Self.precomputed(selection, membership: membership).apply(to: testGraph)
+    case let .combined(lhs, rhs):
+      var result = testGraph
+      result = lhs.apply(to: result)
+      result = rhs.apply(to: result)
+      return result
+    }
+  }
+}
+
+extension Configuration.TestFilter {
+  /// Apply this test filter to a test graph and remove tests that should not be
+  /// included.
+  ///
+  /// - Parameters:
+  ///   - testGraph: The test graph to filter.
+  ///
+  /// - Returns: A copy of `testGraph` with filtered tests replaced with `nil`.
+  func apply(to testGraph: Graph<String, Test?>) -> Graph<String, Test?> {
+    var result = _kind.apply(to: testGraph)
+
+    // After performing the test function, run through one more time and remove
+    // hidden tests. (Note that this property's value is not recursively set on
+    // combined test filters. It is only consulted on the outermost call to
+    // apply(to:), not in _apply(to:).
+    if !includeHiddenTests {
+      result = result.mapValues { test in
+        (test?.isHidden == true) ? nil : test
+      }
+    }
+
+    return result
+  }
+
+  /// Combine this test filter with another one.
```

Comment:
nit: these `combining` functions might make sense in their own `extension`, separate from `apply()` above